include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -30,65 +29,72 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
 		using base_t = asset::IDescriptorSet<const IGPUDescriptorSetLayout>;
 
 	public:
-		IGPUDescriptorSet(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout, core::smart_refctd_ptr<IDescriptorPool>&& pool)
-			: base_t(std::move(_layout)), IBackendObject(std::move(dev)), m_version(0ull), m_pool(std::move(pool))
-		{
-            allocateDescriptors();
-
-            for (auto i = 0u; i < asset::EDT_COUNT; ++i)
-            {
-                // There is no descriptor of such type in the set.
-                if (m_descriptorStorageOffsets.data[i] == ~0u)
-                    continue;
+        IGPUDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout, core::smart_refctd_ptr<IDescriptorPool>&& pool, IDescriptorPool::SDescriptorOffsets&& offsets);
 
-                const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(i);
+        inline uint64_t getVersion() const { return m_version.load(); }
 
-                // Default-construct the core::smart_refctd_ptr<IDescriptor>s because even if the user didn't update the descriptor set with ILogicalDevice::updateDescriptorSet we
-                // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
-                std::uninitialized_default_construct_n(getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
-            }
+        inline IDescriptorPool* getPool() const { return m_pool.get(); }
 
-            const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
-            if (mutableSamplerCount > 0)
-                std::uninitialized_default_construct_n(getMutableSamplerStorage() + m_descriptorStorageOffsets.data[asset::EDT_COUNT], mutableSamplerCount);
+        inline bool isZombie() const
+        {
+            if (m_pool->m_version.load() > m_parentPoolVersion)
+                return true;
+            else
+                return false;
         }
 
-        inline uint64_t getVersion() const { return m_version.load(); }
+	protected:
+        virtual ~IGPUDescriptorSet();
 
-        inline void incrementVersion() { m_version.fetch_add(1ull); }
+	private:
+        friend class ILogicalDevice;
 
-        inline IDescriptorPool* getPool() const { return m_pool.get(); }
+        inline void incrementVersion() { m_version.fetch_add(1ull); }
 
-        inline core::smart_refctd_ptr<asset::IDescriptor>* getAllDescriptors(const asset::E_DESCRIPTOR_TYPE type) const
+        // TODO(achal): Don't know yet if we want to keep these.
+        inline void processWrite(const IGPUDescriptorSet::SWriteDescriptorSet& write)
```

Comment:
hmm

Reply:
also fold `incrementVersion()` into these

Reply:
Also try to detect failures and return a boolean at the end (don't return right way) if there's any fuckups like binding cannot be found, etc.

Reply:
I think all pointers except samplers shoul never be nullptr, and even then the sampler pointers can be reasoned about whether they need to be nullptr or not (Due to layouts or presence in the copy source)